dev/WindowsAppRuntime_BootstrapDLL/MddBootstrap.cpp

```diff
@@ -74,11 +88,61 @@ static std::wstring g_test_ddlmPackagePublisherId;
 STDAPI MddBootstrapInitialize(
     UINT32 majorMinorVersion,
     PCWSTR versionTag,
-    PACKAGE_VERSION minVersion) noexcept try
+    PACKAGE_VERSION minVersion) noexcept
+{
+    return MddBootstrapInitialize2(majorMinorVersion, versionTag, minVersion, MddBootstrapInitializeOptions_None);
+}
+
+STDAPI MddBootstrapInitialize2(
+    UINT32 majorMinorVersion,
+    PCWSTR versionTag,
+    PACKAGE_VERSION minVersion,
+    MddBootstrapInitializeOptions options) noexcept try
 {
     // Dynamic Dependencies Bootstrap API requires a non-packaged process
     LOG_HR_IF(HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED), AppModel::Identity::IsPackagedProcess());
 
+    // Are we already initialized?
+    const auto hr{ _MddBootstrapInitialize(majorMinorVersion, versionTag, minVersion) };
+    if (FAILED(hr))
+    {
+        LOG_IF_FAILED(MddBootstrapInitialize_Log(hr, majorMinorVersion, versionTag, minVersion));
+
+        if (WI_IsFlagSet(options, MddBootstrapInitializeOptions_OnError_DebugBreak) ||
+            (WI_IsFlagSet(options, MddBootstrapInitializeOptions_OnError_DebugBreak_IfDebuggerAttached) && IsDebuggerPresent()) ||
+            IsOptionEnabled(L"MICROSOFT_WINDOWSAPPRUNTIME_BOOTSTRAP_INITIALIZE_DEBUGBREAK"))
+        {
+            DebugBreak();
```

Comment:
this will int3 if only MddBootstrapInitializeOptions_OnError_DebugBreak is set, and invoke Watson, which is more or less what the FailFast support provides.  is that useful?

Reply:
You're suggesting the DebugBreak + FailFast are the same? If a debugger's attached they're not. If no debugger doesn't fail fast do more?

Reply:
I'm questioning the usefulness of MddBootstrapInitializeOptions_OnError_DebugBreak, when running without debugger.  That's a hardcoded int3 that takes the process down, presumably with a JIT debugger (watson) to clean things up.  Doesn't seem like a useful scenario beyond the FailFast option.

I would suggest folding the two DebugBreak options into just MddBootstrapInitializeOptions_OnError_DebugBreak, with IsDebuggerPresent().

Reply:
`DebugBreak if IsDebuggerPresent()=FALSE` isn't the same as `FailFast`. IsDebuggerPresent() tells you if the process is currently under a debugger; it doesn't return `TRUE` if AeDebug is defined.

There's also some logging and telemetry differences for DebugBreak vs FailFast.

Given you have no ability to alter flow here other than these options I'm inclined to keep both `DebugBreak` and `DebugBreak_IfDebuggerPresent`.




More details on how FailFast and DebugBreak are equivalent in concept, but differ in detail...

Both cause the process to terminate if no debugger and AeDebug isn't set. EventLog's slightly different
* ERROR: Exception code = 0xC00000409 vs 0x80000003
* INFORMATION: Event Name = BEX64 vs APPCRASH (this was an x64 exe on an x64 system; BEX64 probably varies for x86 or arm64)
* INFORMATION: P7=address(?), P8=exception code, p9=??? vs P7=exception code, P8=(a different) address(?)

If AeDebug is enabled then the debugger pops up but with very different state
* Fatal process is terminating vs continues on after DebugBreak


Detailed information...

**FailFast**
![image](https://user-images.githubusercontent.com/72222/160464529-f9507514-881a-4520-89dc-262fee229ecd.png)

Error:
There was no match for the specified key in the index.

Error:
Faulting application name: HelloWorldAdvancedCPP.exe, version: 0.0.0.0, time stamp: 0x62313c9d
Faulting module name: KERNELBASE.dll, version: 10.0.19041.1566, time stamp: 0x0833f2d4
Exception code: 0xc0000409
Fault offset: 0x000000000010b362
Faulting process id: 0x19888
Faulting application start time: 0x01d842d22bfca769
Faulting application path: C:\source\repos\projectreunionInternal\BuildOutput\Debug\x64\HelloWorldAdvancedCPP\HelloWorldAdvancedCPP.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report Id: 423f7996-9b19-4a12-95b0-d7b6fad5c24f
Faulting package full name: 
Faulting package-relative application ID: 

Information:
Fault bucket 1462433526648390129, type 5
Event Name: BEX64
Response: Not available
Cab Id: 0

Problem signature:
P1: HelloWorldAdvancedCPP.exe
P2: 0.0.0.0
P3: 62313c9d
P4: KERNELBASE.dll
P5: 10.0.19041.1566
P6: 0833f2d4
P7: 000000000010b362
P8: c0000409
P9: 0000000000000007
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERBCFA.tmp.dmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERBD49.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERBD69.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERBD67.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERBD97.tmp.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_HelloWorldAdvanc_5dadd263235e2195da207dcc877eacdbf82a21f_ce7060d9_5b59924f-5f13-44ed-bcae-556e117bc5f1

Analysis symbol: 
Rechecking for solution: 0
Report Id: 423f7996-9b19-4a12-95b0-d7b6fad5c24f
Report Status: 268435456
Hashed bucket: 9d3a75106516a09a744b9b8b4f7fc1f1
Cab Guid: 0

**DebugBreak**
![image](https://user-images.githubusercontent.com/72222/160464636-c2aa3a5d-c53b-4495-8297-6623006e9a71.png)

Error:
There was no match for the specified key in the index.

Error:
Faulting application name: HelloWorldAdvancedCPP.exe, version: 0.0.0.0, time stamp: 0x62313c9d
Faulting module name: KERNELBASE.dll, version: 10.0.19041.1566, time stamp: 0x0833f2d4
Exception code: 0x80000003
Fault offset: 0x00000000000c90d2
Faulting process id: 0xb6c0
Faulting application start time: 0x01d842d293fc852e
Faulting application path: C:\source\repos\projectreunionInternal\BuildOutput\Debug\x64\HelloWorldAdvancedCPP\HelloWorldAdvancedCPP.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report Id: 0970faf8-13bc-4a83-8680-f34ba58454fc
Faulting package full name: 
Faulting package-relative application ID: 

Information:
Fault bucket 1789497600844845430, type 4
Event Name: APPCRASH
Response: Not available
Cab Id: 0

Problem signature:
P1: HelloWorldAdvancedCPP.exe
P2: 0.0.0.0
P3: 62313c9d
P4: KERNELBASE.dll
P5: 10.0.19041.1566
P6: 0833f2d4
P7: 80000003
P8: 00000000000c90d2
P9: 
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER666F.tmp.dmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER66AE.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER66CE.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER66DC.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER66FC.tmp.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_HelloWorldAdvanc_9f369cebc7cd67252bb2e93dcf8fe3370961c5_ce7060d9_9dea9c30-8594-454c-abb4-92cf7401ffc5

Analysis symbol: 
Rechecking for solution: 0
Report Id: 0970faf8-13bc-4a83-8680-f34ba58454fc
Report Status: 268435456
Hashed bucket: f93bd8b151199ec258d5929638c55176
Cab Guid: 0

FailFast AEDebug 
![image](https://user-images.githubusercontent.com/72222/160466258-2ef25657-ae20-42d3-ba56-8c8712e2f68d.png)

DebugBreak
![image](https://user-images.githubusercontent.com/72222/160466435-0cef8406-9493-4f70-a337-7e9cbda6f6b0.png)

Reply:
I'm not persuaded that DebugBreak (without debugger present) is a useful production option.  A hardcoded int3 (aka 80000003) is typically an indication that code has shipped with an ASSERT - a mistake.  I suppose we can leave it up to developers to decide.